aggregator/aggregator.go

```diff
@@ -157,24 +197,48 @@ func (a *Aggregator) AddOrCreate(key string, ts uint32, quantized uint, value fl
 
 // Flush finalizes and removes aggregations that are due
 func (a *Aggregator) Flush(ts uint) {
+	flushWaiting.Inc(1)
 	flushes.Add()
-	for k, proc := range a.aggregations {
-		if k.ts <= ts {
+	flushWaiting.Dec(1)
+	defer flushes.Done()
+
+	pos := -1 // will track the pos of the last ts position that was successfully processed
+	for i, quantized := range a.tsList {
+		if quantized > ts {
+			break
+		}
+		for key, proc := range a.aggregations[ts] {
 			results, ok := proc.Flush()
 			if ok {
 				if len(results) == 1 {
-					a.out <- []byte(fmt.Sprintf("%s %f %d", k.key, results[0].val, k.ts))
+					a.out <- []byte(fmt.Sprintf("%s %f %d", key, results[0].val, quantized))
+					a.numFlushed.Inc(1)
 				} else {
 					for _, result := range results {
-						a.out <- []byte(fmt.Sprintf("%s.%s %f %d", k.key, result.fcnName, result.val, k.ts))
+						a.out <- []byte(fmt.Sprintf("%s.%s %f %d", key, result.fcnName, result.val, quantized))
+						a.numFlushed.Inc(1)
 					}
 				}
 			}
-			delete(a.aggregations, k)
 		}
+		delete(a.aggregations, ts)
+		pos = i
+	}
+	// now we must delete all the timestamps from the ordered list
+	if pos == -1 {
+		// we didn't process anything, so no action needed
+		return
 	}
+	if pos == len(a.tsList)-1 {
+		// we went through all of them. can just reset the slice
+		a.tsList = a.tsList[:1]
+		return
+	}
+
+	copy(a.tsList[0:], a.tsList[pos+1:])
+	a.tsList = a.tsList[:len(a.tsList)-pos-1]
+
```

Comment:
Looking at this whole block, would it make more sense to do something like:
```
for quantized, ok := a.tsList[0]; ok && quantized < ts {
  // process a.aggregations[quantized]
  delete(a.aggregations, quantized)
  a.tsList = a.tsList[1:]
}
```
Also not sure why the current code uses `a.aggregations[ts]` vs `a.aggregations[quantized]`

Reply:
1) what replay said. don't want to waste slots in the beginning of the slice. will document better
2) `a.aggregations[ts]` this is a bug! will fix . actually will rename ts to cutoff and quantized to ts.
3) re moving the timestamp comparison to the for condition. i think both styles have their proponents, unless someone can link to an established style guide that we can implement, let's focus on things that matter more.

Reply:
1. my understanding is that go actually handles that by copying when required to keep the array reasonable, at least per https://grokbase.com/t/gg/golang-nuts/136crhcaw1/go-nuts-slice-remove-element-memory-question (it's also the approach outlined in https://github.com/golang/go/wiki/SliceTricks#pop-frontshift
2. cool
3. Right now the logic for how we update `a.tsList` is convoluted, and the cleanup is separated from the `delete` from `a.aggregations`.  The intent is to remove that complexity and make the code easier to follow, and remove the processed timestamp from both `a.tsList` and `a.aggregations` at the same time.

Reply:
3/1 fair enough, we can look for ways to simplify.. but `a.tsList = a.tsList[1:]` doesn't seem like a good choice because it'll allocate a new backing array and copy each time we process a `ts`. this generates lots of garbage and overhead (btw wow when did the compiler start doing this). The "example 2" of the first link is basically a more elegant (though also more opaque) way of writing copy statements.  But.. do we want to trade a single copy for O(num-ts-being-processed) copies?  Luckily normally num-ts-being-processed should be very low (typically 1), so that makes it a reasonable decision. but i would rather write them as explicit copies rather than relying than append implementation details.

Reply:
The new backing array isn't allocated when you slice, but rather when you `append`, in the current code you're still going to get the allocation when we append regardless of how we handle things here, so if I'm understanding things correctly it'll be more efficient to use `a.tsList = a.tsList[1:]` here and let golang handle re-allocating the backing array when we append on line 176.  Am I missing something?

Reply:
Oh and btw `quantized, ok := a.tsList[0]` doesn't work on slices.

Reply:
Lots of ways to skin that cat, eg:
```
for len(a.tsList) > 0 && a.tsList[0] < cutoff {
  quantized := a.tsList[0]
```

Reply:
It won't always grow when we append, it depends on how much capacity we have left.
To demonstrate: https://play.golang.org/p/cIOo33SEqff

Reply:
We've gone ahead with this PR because we need it.
These kinds of cleanups are nice to haves but we shouldn't spend too much time on it.